### PR TITLE
Refactor graphql-migrations

### PR DESCRIPTION
**Pull request by craicoverflow**
## Tasks

- [x] Prevent auto deletion of columns and tables
- [x] Add `@drop` directive to indicate the user wants to drop the table or database
- [ ] Use `@drop` directive to drop table/column 
- [ ] Generate/persist migration metadata ???
- [ ] Generate mapping file to map GraphQL field/table to database field/table name.

## Verification

1. Run runtime-example. Migration is called [here](https://github.com/craicoverflow/graphback/pull/2/files#diff-c547ec3122dc95c5210e5d58f7789571R23)
2. Database will be modified to match the input schema.
3. Removing a model/field from the input schema will not remove it from the database

### CODE REVIEW

File: docs/database-schema-migrations.md [hunk 1]
```diff
@@ -30,7 +30,7 @@ Here is an example of how to configure database initialization strategies.
 ```ts
 import * as jsonConfig from '../graphback.json'
 import { schemaText } from './schema';
-import { migrate, UpdateDatabaseIfChanges } from 'graphql-migrations';
+import { migrate, UpdateDatabaseIfChanges } from 'graphql-migrations-bk';
```
Comment (craicoverflow) [inline docs/database-schema-migrations.md]: keeping old migrations for now for backward compatibility.

File: examples/runtime-example/src/runtime.ts [hunk 2]
```diff
@@ -20,9 +20,10 @@ export const createRuntime = async (client: Knex) => {
   const backend = new GraphQLBackendCreator(schemaText, jsonConfig.graphqlCRUD);
   const dbClientProvider = new PgKnexDBDataProvider(client);
 
-  const dbInitialization = new UpdateDatabaseIfChanges(client, jsonConfig.folders.migrations);
-
-  await migrate(schemaText, dbInitialization);
+  const ops = await migrate({
```
Comment (craicoverflow) [inline examples/runtime-example/src/runtime.ts]: API is similar

File: packages/graphql-migrations/src/diff/computeDiff.ts [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+          const toTableIndex = addTableQueue.indexOf(toTable)
+          // A new table shouldn't be added
+          addTableQueue.splice(toTableIndex, 1)
+          break
+        }
+      }
+
+      // Drop table
+      if (removed) {
+        console.warn('To drop a table use the @drop directive')
```
Comment (craicoverflow) [inline packages/graphql-migrations/src/diff/computeDiff.ts]: Tables will not be dropped automatically.

File: packages/graphql-migrations/src/diff/computeDiff.ts [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+
+        // Rename table
+        const { annotations, directives } = toTable
+        if (annotations && annotations.oldNames && annotations.oldNames.includes(fromTable.name)) {
+          removed = false
+          this.renameTable(fromTable, toTable)
+        }
+
+        // TODO: Drop table with drop directive
+        // if (directives.drop) {
```
Comment (craicoverflow) [inline packages/graphql-migrations/src/diff/computeDiff.ts]: This is a WIP. We can use directives to force-drop the table.

File: examples/runtime-example/src/runtime.ts (hunk 2 above)
Comment (wtrocki) [reply examples/runtime-example/src/runtime.ts]: What is schema text here ? Model? Are we stripping out migration directives

File: packages/graphback-cli/src/commands/db.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import chalk from 'chalk';
-import { DropCreateDatabaseAlways } from 'graphql-migrations';
+import { DropCreateDatabaseAlways } from 'graphql-migrations-bk';
```
Comment (wtrocki) [inline packages/graphback-cli/src/commands/db.ts]: IMHO we do not need backward compatbility

Comment (craicoverflow) [reply packages/graphback-cli/src/commands/db.ts]: I don't plan to keep it. I didn't want to unnecessarily disrupt the CLI for what is a spike. I am focusing on runtime for now and so this was the easiest thing to do to not break things.

File: examples/runtime-example/src/runtime.ts (hunk 2 above)
Comment (craicoverflow) [reply examples/runtime-example/src/runtime.ts]: Input model. Why would we strip out migration directives when we need them to run the migration?

**Comment by wtrocki** [comment]
@craicoverflow Can you update description. Mainly - what is the target. What are the current challenges etc.

**Comment by craicoverflow** [comment]
> @craicoverflow Can you update description. Mainly - what is the target. What are the current challenges etc.

I will open PR on Aerogear project.

**Pull request closed without merging**